ir/memory.cpp

```diff
@@ -815,11 +811,29 @@ expr Pointer::blockValRefined(const Pointer &other) const {
                 val.ptrByteoffset() == val2.ptrByteoffset() &&
                 val.ptr().refined(val2.ptr());
   }
+  // allow null ptr <-> zero
+  // If ptr points to a null block with ofs, its integer address is compared.
+  // Relevant transformations:
+  //   store ty* null  <-> memset i8 0
+  //   store ty* undef  -> memset i8 undef
+  expr intval;
+  if (val.isValid()) {
+    expr ofs = val.ptr().getOffset();
+    expr shift_amount = val.ptrByteoffset() *
+                        expr::mkUInt(bits_byte, val.ptrByteoffset().bits());
+    intval = ofs.lshr(shift_amount.zextOrTrunc(ofs.bits()))
+                .zextOrTrunc(bits_byte);
+  }
+  // val is (null, ofs), val2 is ofs
+  expr refine_ptrint = val.ptr().isNull(false) && !val2.isPoison(false) &&
+                       intval == val2.nonptrValue();
+  // val is 0, val2 is nullptr
+  expr refine_intptr =
+    !val2.isPoison(false) && val.nonptrValue() == 0 && val2.ptr().isNull();
   return val.isPoison() ||
-         expr::mkIf(is_ptr == is_ptr2,
-                    expr::mkIf(is_ptr, ptr_cnstr, int_cnstr),
-                    // allow null ptr <-> zero
-                    val.isZero() && !val2.isPoison() && val2.isZero());
+         expr::mkIf(is_ptr,
```

Comment:
I feel the poison constraint for val2 needs to be hoisted. Right now it appears 4x times. Not good.
Also, don't use `isPoison` as it brings a big formula for no reason.

Reply:
Updated to use nonptrNonpoison instead.
Is hoisting the poison check for val2 still needed? It is simpler, compared to the before. And to me hoisting it seemed nontrivial...